### PR TITLE
[js/web] JSEP LayerNormalization and InstanceNormalizations kernels

### PR DESCRIPTION
**Pull request by dakenf**
### Description
Added two kernels for Layer and Instance norm

Also added maximum limits for `maxBufferSize` when requesting GPU device as by default it's limited to 256mb and it fails allocating 600mb buffer while running fp32 StableDiffusion weights.

### Motivation and Context
These two are used in StableDiffusion and many other networks

### CODE REVIEW

File: onnxruntime/core/providers/js/operators/layer_norm.h
Comment (satyajandhyala) [inline onnxruntime/core/providers/js/operators/layer_norm.h:15]: GetAttr function requires that the attribute is specified. According to the [documentation](https://github.com/onnx/onnx/blob/main/docs/Operators.md#layernormalization) the attributes, axis, epsilong and stash_type are optional, i.e. have default values, -1, 1e-05, and 1 respectively. Please use GetAttrOrDefault.

File: onnxruntime/core/providers/js/operators/instance_norm.h
Comment (satyajandhyala) [inline onnxruntime/core/providers/js/operators/instance_norm.h:14]: Use GetAttrOrDefault instead of GetAttr which requires that the attribute is set.

**Comment by satyajandhyala** [comment]
Can we verify/enable some of these tests if not all?
https://github.com/microsoft/onnxruntime/blob/5d17bcd776f326506360187b48808c2c005dbc94/js/web/test/suite-test-list.jsonc#L609C1-L646C53

File: onnxruntime/core/providers/js/operators/instance_norm.h
Comment (fs-eire) [inline onnxruntime/core/providers/js/operators/instance_norm.h:11]: we have a macro to replace the whole class definition: `JSEP_CLASS_IMPL_ATTRIBUTE_FLOAT_DEFAULT` (defined in js_kernel.h)

File: onnxruntime/core/providers/js/operators/layer_norm.h
Comment (fs-eire) [inline onnxruntime/core/providers/js/operators/layer_norm.h:22]: should use int32_t instead of uint32_t for axis.

from spec:

axis - INT (default is '-1'): The first normalization dimension. If rank(X) is r, axis’ allowed range is [-r, r]. Negative value means counting dimensions from the back.

File: onnxruntime/core/providers/js/operators/layer_norm.h
Comment (fs-eire) [inline onnxruntime/core/providers/js/operators/layer_norm.h:23]: from the spec, there is another attribute `stash_type`. I guess this is only for tranning? if we ignore this attribute intentionally, please put a comment there explaining.

File: onnxruntime/core/providers/js/js_execution_provider.cc
Comment (fs-eire) [inline onnxruntime/core/providers/js/js_execution_provider.cc:438]: if we want to support the NHWC version of InstanceNormalization as well, we need to pass the format to JavaScript so that in JavaScript we can know whether it's NCHW or NHWC.

see onnxruntime\core\providers\js\operators\conv.h for how `"format"` is being passed.

File: onnxruntime/core/providers/js/operators/layer_norm.h
Comment (dakenf) [reply onnxruntime/core/providers/js/operators/layer_norm.h:15]: Ok, but in CPU provider OP `onnxruntime/core/providers/cpu/nn/layer_norm_impl.cc` it does not use default values. Does it get the default ones in some another way for CPU?
```
LayerNormImpl::LayerNormImpl(const OpKernelInfo& op_kernel_info, bool simplified, bool contrib_op)
    : OpKernel(op_kernel_info), simplified_{simplified}, contrib_op_{contrib_op} {
  ORT_ENFORCE(op_kernel_info.GetAttr("axis", &axis_).IsOK());
  ORT_ENFORCE(op_kernel_info.GetAttr<float>("epsilon", &epsilon_).IsOK());
}
```

File: onnxruntime/core/providers/js/operators/layer_norm.h
Comment (dakenf) [reply onnxruntime/core/providers/js/operators/layer_norm.h:23]: I've copied behavior of CPU OP from `onnxruntime/core/providers/cpu/nn/layer_norm_impl.h` and it did not have this attribute 🤔

**Comment by dakenf** [comment]
> Can we verify/enable some of these tests if not all?

I've enabled all instance norm tests (there were just two) and layer norm except "expanded" ones which don't actually test the OP and fail for some reason. Maybe ReduceMean results are incorrect? Here is the onnx file for one of expanded tests:
image

File: onnxruntime/core/providers/js/js_execution_provider.cc
Comment (dakenf) [reply onnxruntime/core/providers/js/js_execution_provider.cc:438]: I've added a condition and a different shader code for NHWC

**Comment by dakenf** [comment]
@fs-eire @satyajandhyala can you please advise me about a weird issue i get with LayerNorm?
In the spec and for tests it requires three outputs: result, mean data and inv std data. And it works fine.
But when i run it with StableDiffusion text encoder or unet it throws an error:
```
Writable storage buffer binding aliasing found between [BindGroup] set at bind group index 0, binding index 4, and [BindGroup] set at bind group index 0, binding index 5, with overlapping ranges (offset: 0, size: 1280) and (offset: 0, size: 1280) in [Buffer].
 - While encoding [ComputePassEncoder].DispatchWorkgroups(8, 1, 1).
 ```
So if i comment out the last binding it works perfectly fine but tests fail as they expect three outputs
```javascript
@group(0) @binding(0) var<storage, read> x : array<${dataType}>;
  @group(0) @binding(1) var<storage, read> scale : array<${dataType}>;
  @group(0) @binding(2) var<storage, read> bias : array<${dataType}>;
  @group(0) @binding(3) var<storage, read_write> output : array<${dataType}>;
  @group(0) @binding(4) var<storage, read_write> meanDataOutput : array<${dataType}>;
  // COMMENT OUT THIS @group(0) @binding(5) var<storage, read_write> invStdOutput : array<${dataType}>;

  ${shaderHelper.mainStart(workgroupLimits)}
   ... shader code

    meanDataOutput[global_idx] = mean;
    invStdOutput[global_idx] = 1 / meanSquare;
  }`;
        return {
            ...metadata,
            outputs: [
                {dims: outputShape, dataType: inputs[0].dataType, gpuDataType: GpuDataType.default},
                {dims: meanInvStdDevDim, dataType: inputs[0].dataType, gpuDataType: GpuDataType.default},
                 // COMMENT OUT THIS {dims: meanInvStdDevDim, dataType: inputs[0].dataType, gpuDataType: GpuDataType.default},
            ],
            getShaderSource,
            dispatchGroup: () => (dispatchGroup)
        };
    };
```

-------------
And i have another unrelated issue. After adding Gemm for opset 13 (currenty in JSEP it has a kernel only for 11 and falls back to CPU) i get this error:
```
[Buffer] usage (BufferUsage::(Storage|BufferUsage::80000000)) includes writable usage and another usage in the same synchronization scope.
 - While validating compute pass usage.
```
I guess it hits max buffer count/sizes. Is there an easy way to dispose unused buffers during the execution? As it uses more than 8gb of video memory. 3gb for UNET weights and allocates more than 5 during OrtRun.

**Comment by fs-eire** [comment]
The output count needs to match the actual number of the node's outputs in the graph. I am not sure if a `context` can get the outputs number, if not we need to add this.

**Comment by satyajandhyala** [comment]
> @fs-eire @satyajandhyala can you please advise me about a weird issue i get with LayerNorm? In the spec and for tests it requires three outputs: result, mean data and inv std data. And it works fine. But when i run it with StableDiffusion text encoder or unet it throws an error:
> 
> ```
> Writable storage buffer binding aliasing found between [BindGroup] set at bind group index 0, binding index 4, and [BindGroup] set at bind group index 0, binding index 5, with overlapping ranges (offset: 0, size: 1280) and (offset: 0, size: 1280) in [Buffer].
>  - While encoding [ComputePassEncoder].DispatchWorkgroups(8, 1, 1).
> ```
> 
> So if i comment out the last binding it works perfectly fine but tests fail as they expect three outputs
> 
> ```js
> @group(0) @binding(0) var<storage, read> x : array<${dataType}>;
>   @group(0) @binding(1) var<storage, read> scale : array<${dataType}>;
>   @group(0) @binding(2) var<storage, read> bias : array<${dataType}>;
>   @group(0) @binding(3) var<storage, read_write> output : array<${dataType}>;
>   @group(0) @binding(4) var<storage, read_write> meanDataOutput : array<${dataType}>;
>   // COMMENT OUT THIS @group(0) @binding(5) var<storage, read_write> invStdOutput : array<${dataType}>;
> 
>   ${shaderHelper.mainStart(workgroupLimits)}
>    ... shader code
> 
>     meanDataOutput[global_idx] = mean;
>     invStdOutput[global_idx] = 1 / meanSquare;
>   }`;
>         return {
>             ...metadata,
>             outputs: [
>                 {dims: outputShape, dataType: inputs[0].dataType, gpuDataType: GpuDataType.default},
>                 {dims: meanInvStdDevDim, dataType: inputs[0].dataType, gpuDataType: GpuDataType.default},
>                  // COMMENT OUT THIS {dims: meanInvStdDevDim, dataType: inputs[0].dataType, gpuDataType: GpuDataType.default},
>             ],
>             getShaderSource,
>             dispatchGroup: () => (dispatchGroup)
>         };
>     };
> ```
> 
> And i have another unrelated issue. After adding Gemm for opset 13 (currenty in JSEP it has a kernel only for 11 and falls back to CPU) i get this error:
> 
> ```
> [Buffer] usage (BufferUsage::(Storage|BufferUsage::80000000)) includes writable usage and another usage in the same synchronization scope.
>  - While validating compute pass usage.
> ```
> 
> I guess it hits max buffer count/sizes. Is there an easy way to dispose unused buffers during the execution? As it uses more than 8gb of video memory. 3gb for UNET weights and allocates more than 5 during OrtRun.

Does it help specifying the size in addition to dataType for x, scale, bias, output, etc.?

**Comment by dakenf** [comment]
> Does it help specifying the actual size, if possible, in addition to dataType for x, scale, bias, output, etc.?

The suggestion above helped. I've added OutputCount to serialized kernel context and it started to work fine

**Comment by fs-eire** [comment]
I have a question regarding the optional inputs/outputs of an operator. take `LayerNormalization` for example - is it possible that a graph contains an LayerNormalization node whose output[0] and output[2] exist, but output[1] is omitted? If this situation may happen, then defining 'outputCount' may not be sufficient (it'll be ambiguous because if outputCount==2, it may mean either output[0]+output[1] or output[0]+output[2])

**Comment by dakenf** [comment]
> I have a question regarding the optional inputs/outputs of an operator. take `LayerNormalization` for example - is it possible that a graph contains an LayerNormalization node whose output[0] and output[2] exist, but output[1] is omitted? If this situation may happen, then defining 'outputCount' may not be sufficient (it'll be ambiguous because if outputCount==2, it may mean either output[0]+output[1] or output[0]+output[2])

From what I see in the code, if output[1] does not exist, node will have three outputs but for the second `context->OutputType(1)` will return `nullptr`. But I might be wrong. CPU execution provider does not have any checks, it always has three outputs. 

Also, `bias` input for LayerNorm is optional, so i will add a check for it

**Review by guschmue — APPROVED** [review]

**Comment by dakenf** [comment]
@fs-eire @satyajandhyala @guschmue can I ask you a favor? Maybe you know someone from MS Edge team who can review my V8 pull request https://chromium-review.googlesource.com/c/v8/v8/+/4742982

It allows multi-threading for 64bit webassembly. By the time it gets to browsers, I will finish and clean up my draft wasm64 pull request here and ONNX can be the first runtime able to run big models in the browser with acceptable performance. All required emscripten compiler pull requests are already merged and will be in the next release

**Comment by guschmue** [comment]
yes, I can help with that. Let me forward it.

**Comment by guschmue** [comment]
ok, some reviewers got added to your chromium PR - hope that works out.

**Comment by dakenf** [comment]
Pefect, thanks. If I will be able to fix memory.grow after 4gb then there will be no reason to use 32bit version with JSEP. Right now the only way to use more memory is to specify fixed size with INITIAL_MEMORY linking flag

**Comment by guschmue** [comment]
yes. I'm sure there will be some more complications on the way, but it will be needed now that webgpu is there and shows the possibilities to run things like sd or llm's in the browser.

File: onnxruntime/core/providers/js/operators/instance_norm.h
Comment (satyajandhyala) [reply onnxruntime/core/providers/js/operators/instance_norm.h:12]: The macro is not applicable in this case anymore because the class has defines two attributes, one float and the other int32_t.

File: onnxruntime/core/providers/js/operators/layer_norm.h
Comment (satyajandhyala) [reply onnxruntime/core/providers/js/operators/layer_norm.h]: Let us go by the documentation and use GetAttrOrDefault.

Comment (dakenf) [reply onnxruntime/core/providers/js/operators/layer_norm.h]: Yup. I've updated the code to use GetAttrOrDefault

**Comment by satyajandhyala** [comment]
@dakenf 
The latest change to the webgpu-operator.md document removed some comments existed previously. I am not sure whether that it intentional or not. We don't edit this file directly. It should be generated using `npm run build:doc` in onnxruntime/js/web folder. The comments get inserted by generate-webgpu-operator-md.ts. After making any necessary changes to this TypeScript file you need to run `npm run prepare` to compile/generate JavaScript file.

File: js/web/docs/webgpu-operators.md
Comment (satyajandhyala) [inline js/web/docs/webgpu-operators.md:28]: Removing this comment may be unintentional.

File: js/web/docs/webgpu-operators.md
Comment (satyajandhyala) [inline js/web/docs/webgpu-operators.md:65]: Same here

**Comment by satyajandhyala** [comment]
@dakenf Some unresolved/valid comments are blocking merging this PR. Please address these comments so that this PR can be merged.

File: js/web/docs/webgpu-operators.md
Comment (dakenf) [reply js/web/docs/webgpu-operators.md:65]: I did not update it manually, just used `npm run build:doc` as otherwise the pipeline does not pass

Comment (dakenf) [reply js/web/docs/webgpu-operators.md:65]: image

Comment (satyajandhyala) [reply js/web/docs/webgpu-operators.md:65]: did you run `npm run prepare` command in onnxruntime\js\web directory?

Comment (satyajandhyala) [reply js/web/docs/webgpu-operators.md:65]: Please try running `npm run prepare` in onnxruntime\js\web folder also. It seems you ran it in onnxruntime\js I am not sure it is supposed to run in the subdirectories.

Comment (dakenf) [reply js/web/docs/webgpu-operators.md:65]: Does not help. It started to remove comments recently, after I've merged main branch. Before recent updates it worked as expected
image

Comment (satyajandhyala) [reply js/web/docs/webgpu-operators.md:65]: I have seen this error a while ago. I think running `npm ci` under onnxruntime\js, onnxruntime\js\common, and onnxruntime\js\web directories as mentioned [here](https://gist.github.com/fs-eire/a55b2c7e10a6864b9602c279b8b75dce) might help

Comment (dakenf) [reply js/web/docs/webgpu-operators.md]: Yeah, that helped, thanks

File: onnxruntime/core/providers/js/operators/layer_norm.h
Comment (satyajandhyala) [reply onnxruntime/core/providers/js/operators/layer_norm.h]: @dakenf When you get a chance please fix this last pending comment.

**Comment by satyajandhyala** [comment]
@dakenf Click on "view" link in "Unresolved conversations" to go to the unresolved comment directly.

**Comment by dakenf** [comment]
> @dakenf Click on "view" link in "Unresolved conversations" to go to the unresolved comment directly.

I think it's just github interface lagging because i get this when i click on it
image
But the actual code is already fixed if you go here (with int32_t and axis=-1) https://github.com/microsoft/onnxruntime/pull/16830/files
image
Or am i looking in the wrong place?

**Comment by satyajandhyala** [comment]
@dakenf Thanks for clarifying. I marked the conversation resolved.

**Review by satyajandhyala — APPROVED** [review]

**Review by guschmue — APPROVED** [review]

**Comment by guschmue** [comment]
going to take this for a quick test before I merge it

**Pull request merged**
